src/vm/threadsuspend.cpp

```diff
@@ -8231,6 +8247,126 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
+// one of the following:
+//
+//     - If the thread is in native code or preemptive GC is not disabled, there's
+//       nothing to do, so we return.
+//
+//     - If the thread is in interruptible managed code, we will push a frame that
+//       has information about the context that was interrupted and then switch to
+//       premptive GC mode so that the pending GC can proceed, and then switch back.
+//
+//     - If the thread is in uninterruptible managed code, we will patch the return
+//       address to take the thread to the appropriate stub (based on the return 
+//       type of the method) which will then handle preparing the thread for GC.
+//
+void PALAPI HandleGCSuspensionForInterruptedThread(CONTEXT *interruptedContext)
+{
+    Thread *pThread = GetThread();
+
+    if (pThread->PreemptiveGCDisabled() != TRUE)
+        return;
+
+    PCODE ip = GetIP(interruptedContext);
+
+    if (ExecutionManager::IsManagedCode(ip) != TRUE)
+        return;
+
+    Thread::WorkingOnThreadContextHolder workingOnThreadContext(pThread);
+    if (!workingOnThreadContext.Acquired())
+        return;
+
+    EECodeInfo codeInfo(ip);
+    if (!codeInfo.IsValid())
+        return;
+
+    DWORD addrOffset = codeInfo.GetRelOffset();
+
+    ICodeManager *pEECM = codeInfo.GetCodeManager();
+    _ASSERTE(pEECM != NULL);
+
+    bool isAtSafePoint = pEECM->IsGcSafe(&codeInfo, addrOffset);
+    if (isAtSafePoint)
+    {
+        // If the thread is at a GC safe point, push a RedirectedThreadFrame with
+        // the interrupted context and pulse the GC mode so that GC can proceed.
+        FrameWithCookie<RedirectedThreadFrame> frame(interruptedContext);
+        pThread->SetSavedRedirectContext(NULL);
+
+        frame.Push(pThread);
+
+        pThread->PulseGCMode();
+
+        frame.Pop(pThread);
+    }
+    else
+    {
+        // The thread is in non-interruptible code.
+        ExecutionState executionState;
+        StackWalkAction action;
+        REGDISPLAY regDisplay;
+        pThread->InitRegDisplay(&regDisplay, interruptedContext, true /* validContext */);
+
+        if (!pThread->IsSafeToInjectThreadAbort(interruptedContext))
+            return;
+
+        // Use StackWalkFramesEx to find the location of the return address. This will locate the
+        // return address by checking relative to the caller frame's SP, which is preferable to
+        // checking next to the current RBP because we may have interrupted the function prior to
+        // the point where RBP is updated.
+        action = pThread->StackWalkFramesEx(
+            &regDisplay,
+            SWCB_GetExecutionState,
+            &executionState,
+            QUICKUNWIND | DISABLE_MISSING_FRAME_DETECTION | ALLOW_ASYNC_STACK_WALK);
+
+        if (action != SWA_ABORT || !executionState.m_IsJIT)
+            return;
+
+        if (executionState.m_ppvRetAddrPtr == NULL)
+            return;
+
+        ENABLE_FORBID_GC_LOADER_USE_IN_THIS_SCOPE();
```

Comment:
Sure, will add a comment. I believe this macro doesn't actually impact anything in release mode. Its purpose is to turn off the GC_TRIGGERS/THROWS/INJECT_FAULT contract in LoadTypeHandle in cases where the caller won't try to load a type. I included it here because it's done in the Windows code.